### PR TITLE
Added entrypoint hooks for your own custom scripts

### PR DESCRIPTION
**Pull request by dvaerum**
I what looking around to see if there was anyway to execute custom script(s) when Nextcloud is installed/initiated the first time or whenever it is upgraded.

Sadly, this feature doesn't exist, but I can see there has been talked about it in https://github.com/nextcloud/docker/issues/820#issuecomment-860930549, and I can also see that there are 2 pull requests for adding an environment variable tell the container to install extra apps doing Nextcloud installation/initialization.

- https://github.com/nextcloud/docker/pull/1259
- https://github.com/nextcloud/docker/pull/1260

This entrypoint hooks for custom scripts would allow for people to easily add these features themselves, plus other features like "add support for oauth in docker image" (https://github.com/nextcloud/docker/pull/1023)

I am open for feedback because I think this would give a lot of flexibility with the Nextcloud container :grin:

### CODE REVIEW

File: docker-entrypoint.sh [hunk 1]
```diff
@@ -19,6 +19,33 @@ run_as() {
...
+run_path() {
+    local hook_folder_path="/docker-entrypoint-hooks.d/$1"
+    local return_code=0
+
+    echo "Searching for scripts to run in the folder: ${hook_folder_path}"
+
+    (
+        cd "${hook_folder_path}"
+        find . -type f -print | while read script_file_path; do
+            if ! [ -x "${script_file_path}" ] && [ -f "${script_file_path}" ]; then
```
[the remark this answers is not in the document]
Comment (dvaerum) [reply docker-entrypoint.sh:32]: The "strong reason" for allowing anything executable was to have a way to make an easy to customizable setup in your Nextcloud setup.
Without having to make pull request for every single feature which you would like to added to nextcloud-docker and at the same time lower the complexity.

This way you can have a script there install apps the first time you start your docker container and have another script which configure these apps.
One example could be Onlyoffice, you only need to install and configure it the first time you start Nextcloud and after that it is not needed again.

And to your example. You could have a script which  would do something to the postgres database and there you go magic, no need for any new pull request.
Just in general it would make having a custom test environment, for test if something works or not much easier, and at the end, you could just blow everything away again. No need to for manually configurations of anything.

Also, if you don't volume mount a folder into the container with any executable files it will not run any extra files, just the standard stuff.

Hope this answered your question 🙂

[the remark this answers is not in the document]
Comment (dvaerum) [reply docker-entrypoint.sh:32]: aaarh, now I understand what you are asking :sweat_smile: Well, my logic was that if someone wanted to run some static binary file, or maybe php-scripts they still would be able 🙂
But if that is what they want, they can just use a shell script to start it (if they want to try to go that way), so if you think it will make it easier to maintain by having it be shell scripts, I am happy to make that change :grin:

[the remark this answers is not in the document]
Comment (dvaerum) [reply docker-entrypoint.sh:32]: Hey @J0WI, I have now done so that only files ending in `.sh` and have the executable flag, will be executed.
I also, did so that the files are sorted before execution because I realized that I had forgotten to do that :sweat_smile:

File: docker-entrypoint.sh [hunk 2]
```diff
@@ -19,6 +19,35 @@ run_as() {
...
+# Execute all executable files in a given directory in alphanumeric order
+run_path() {
+    local hook_folder_path="/docker-entrypoint-hooks.d/$1"
+    local return_code=0
+
+    echo "=> Searching for scripts (*.sh) to run in the folder: ${hook_folder_path}"
+
+    (
+        cd "${hook_folder_path}"
+        find . -type f -iname '*.sh' -print | sort | while read script_file_path; do
```
[the remark this answers is not in the document]
Comment (dvaerum) [reply docker-entrypoint.sh:31]: The idea was that people could make folders, as a way to categorize their scripts, and `find` would be able to locate them 🙂

But if you don't like the idea, I have no problem changing it to
```
cd "${hook_folder_path}"
for f in ./*.sh; do
```
Because I already use `cd`, I don't need `$hook_folder_path/` and the `*` (asterisk) does not work if it is wrapped in `"` (double quote)

File: README.md [hunk 3]
```diff
@@ -202,6 +202,36 @@ To customize other PHP limits you can simply change the following variables:
...
+
+There are 5 hooks
+
+- `pre-installation` Executed before the Nextcloud is installed/initiated
+- `post-installation` Executed after the Nextcloud is installed/initiated
+- `pre-upgrade` Executed before the Nextcloud is upgraded
+- `post-upgrade` Executed after the Nextcloud is upgraded
+- `before-starting` Executed before the Nextcloud is started
+
+To use the hooks triggered by the entrypoint script... either added the script(s) to one/more of the hook folders, which are located at the path `/docker-entrypoint-hooks.d` in the container. It is also possible to use volume(s) to insert the script into the container.
```
[the remark this answers is not in the document]
Comment (dvaerum) [reply README.md:215]: I was to indicate a break before summarizing a list of things, but I removed it and mad a list in Markdown, seems more obvious now that I look at it again

File: README.md [hunk 4]
```diff
@@ -202,6 +202,36 @@ To customize other PHP limits you can simply change the following variables:
...
+
+**Remember:** Only the script(s) markted with the executable flag are executed.
```
[the remark this answers is not in the document]
Comment (dvaerum) [reply README.md]: I have added the hint here :+1: 

https://github.com/nextcloud/docker/pull/1964/commits/d5e66a5b3e0aa14e7f6cef25691b4a59b8df433c#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R219

File: docker-entrypoint.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (dvaerum) [reply docker-entrypoint.sh:32]: I will consider this resolved, unless there are more comments 😉

File: docker-entrypoint.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (dvaerum) [reply docker-entrypoint.sh]: Hey @J0WI

I have done as you requested
- `cd` is removed
- `find` is replace with `for script_file_path in "${hook_folder_path}/"*.sh; do`

I also updated the README.md file to specify that scripts in sub-folders of the hook-folders will not be executed. And, when running a script it will also print the current working directory (I assume this info could be helpful)

**Comment by dvaerum** [comment]
> The build fails: `sh: 1: /docker-entrypoint-hooks.d/pre-installation/*.sh: not found`

aaahhh, that is true, I forgot that this can be a problem. If there are no files in the folder for filling the glob-criteria (`*.sh`) one gets this error, one reasons to use find.

but, I will add something like this to test if the folder is empty, when I get home from work or one of their days :grin: 
```bash
[ -z "$(ls -A test_folder)" ] && echo "Empty" || echo "Not Empty"
```

**Comment by dvaerum** [comment]
Hey @J0WI, I have fixed the bug, hopefully there are no more stupid mistake :sweat_smile: will you approve the workflow? :grin:

**Comment by Nils98Ar** [comment]
Thank you for this feature!

The log says:
`==> The script "/docker-entrypoint-hooks.d/post-installation/bootstrap.sh" in the folder "/docker-entrypoint-hooks.d/post-installation" was skipping, because it didn't have the executable flag`

I have this in my Dockerfile:
```
COPY ./docker-entrypoint-hooks.d/post-installation/bootstrap.sh /docker-entrypoint-hooks.d/post-installation
RUN chmod +x /docker-entrypoint-hooks.d/post-installation/bootstrap.sh && chown www-data:root /docker-entrypoint-hooks.d/post-installation/bootstrap.sh
```

Any idea?

**Comment by dvaerum** [comment]
> Thank you for this feature!
> 
> The log says: `==> The script "/docker-entrypoint-hooks.d/post-installation/bootstrap.sh" in the folder "/docker-entrypoint-hooks.d/post-installation" was skipping, because it didn't have the executable flag`
> 
> I have this in my Dockerfile:
> 
> ```
> COPY ./docker-entrypoint-hooks.d/post-installation/bootstrap.sh /docker-entrypoint-hooks.d/post-installation
> RUN chmod +x /docker-entrypoint-hooks.d/post-installation/bootstrap.sh && chown www-data:root /docker-entrypoint-hooks.d/post-installation/bootstrap.sh
> ```
> 
> Any idea?

🤯 that is a good question 🤔 From what I can see, what you do should work.

If you make a zip-file with your environment and the commands you run, I will debug it. Course of you have found a bug I want to fix it

**Comment by Nils98Ar** [comment]
Beginners mistake: I had `imagePullPolicy: IfNotPresent` configured for the container in K8S 🙂

This works as well:
`COPY --chmod=775 --chown=www-data:root ./docker-entrypoint-hooks.d/post-installation/bootstrap.sh /docker-entrypoint-hooks.d/post-installation`.

**Comment by CheariX** [comment]
I'm trying to add some packages (`inotify-tools`) to my installation. Since I do not want to built my own Docker image, i thought it would be a good idea to put the `apt install` code into the `/docker-entrypoint-hooks.d/before-starting` folder.

However, this is not working:

```
nextcloud_web  | E: Could not open lock file /var/lib/dpkg/lock-frontend - open (13: Permission denied)
nextcloud_web  | E: Unable to acquire the dpkg frontend lock (/var/lib/dpkg/lock-frontend), are you root?
```
It seems that the scripts are executed with the `www-data` user since I am using the `docker-compose.yml` from this repository with the Apache setup.
Can anyone point me to how this is supposed to work? The README.md is only providing very rough information on this.

**Comment by dvaerum** [comment]
First of, this was never meant to solve the problem of installing missing packages, this was meant to configure Nextcloud by running commands commands like ‘php occ app:install <APP>’

In my opinion you should make a new image, I don’t know how you start your image, but it is pretty straightforward to make small adjustments to the image using docker-compose.

But, if you want to make the changes/Install the  package when the container start, you can change the entrypoint, to be something like this ‘apt update && apt install —yes <package> && /entrypoint.sh‘. That would give you a root shell to install your package(s).

Hopes this helps/point you in the right direction.

**Comment by CheariX** [comment]
Thanks for the clarification. Your idea with the entryhooks works great for me.
Just for the record, the `docker-compose.yml` entry is:

```
entrypoint: ["/bin/bash", "-c" , "apt-get update && apt-get install curl -y && /entrypoint.sh apache2-foreground"]
```
